aip/0194.md

```diff
@@ -0,0 +1,88 @@
+---
+aip:
+  id: 194
+  state: reviewing
+  created: 2019-08-16
+permalink: /194
+redirect_from:
+  - /0194
+---
+
+# Client-side retries
+
+This document aims to lay out guidance about which requests to retry, and for
+which error codes.
+
+## Guidance
+
+Clients **should** retry requests for which repeated runs would not cause
+unintended state changes, which are non-transactional, and which are either
+unary or server streaming.
+
+Clients **should not** retry transactional requests; instead these requests
+**should** have user-level retry logic that retries the entire transaction
+block from the start.
+
+Clients **should not** retry requests in which repeated runs would cause
+unintended state changes.
```

Comment:
How do we treat billing here? If I retry a request to video intelligence, for example, that might be expensive, but the request is otherwise idempotent.

We may not have to decide this. My gut instinct is that requests that cost extremely little (which is most of them) are fine to retry, but I would want to be a bit careful if the request was expensive.

Reply:
I think that for the things we are recommending retry - INTERNAL and UNAVAILABLE - there is an argument that no billing should occur. But, you're right, it's a good axis to think about. I'll take your easy-out of not dealing with it here, and let's see if it comes up during review with the individual service owners.